Add ExtensionImages directive to form overlays

bluca opened the pull request:
Add support for overlaying the image for services on top of their
root fs.

bluca commented:
Split out from https://github.com/systemd/systemd/pull/16889 - turns out I found this quite useful by itself, for MountImages, and I am currently using it successfully. PTAL!

bluca commented:
Mmmh the autopkgtest suite didn't run at all for some reason, probably testbed issues - force pushing with no change to trigger it

poettering commented on src/core/namespace.c:
```diff
@@ -1019,6 +1028,30 @@ static int mount_images(const MountEntry *m) {
+                return -ENOMEM;
+
+        /* We only support read-only overlays, so no upper nor work directories */
+        options = strjoina("lowerdir=", escaped_source, ":", where);
```
`where` needs to be escaped too, no?

bluca replied:
yes indeed, it goes as an option too, good catch

poettering commented on man/systemd.exec.xml:
```diff
@@ -381,6 +381,14 @@
+        For example, if the root has <filename>/opt/</filename> and <filename>/var/foo</filename>, and the
+        image has <filename>/opt/bar</filename> and <filename>/var/baz</filename>, the end result will be
+        that <filename>/opt/</filename> from the image is bind-mounted on <filename>/opt/</filename> on the root,
+        and <filename>/var</filename> from the root and the image will both be part of an OverlayFS mount.</para>
```
we typically suffix all dirs with / to indicate they are a dir. is /var/foo and /var/baz supposed to be a dir? (if not, it might be good to mention that somewhow

bluca replied:
they were files in the example - changed to specify files/directories explicitly

poettering commented on man/systemd.exec.xml:
```diff
@@ -381,6 +381,14 @@
+        image has <filename>/opt/bar</filename> and <filename>/var/baz</filename>, the end result will be
+        that <filename>/opt/</filename> from the image is bind-mounted on <filename>/opt/</filename> on the root,
+        and <filename>/var</filename> from the root and the image will both be part of an OverlayFS mount.</para>
+
```
hmm, it's not clear to me, why the destination being unspecified should mean "overlayfs"? i.e. why wouldn't i want to use overlayfs logic with a specific specified destination path?

i.e. shouldn't this be a new option OverlayMount= or so?

bluca replied:
The overlay logic to me makes sense root-on-root, so no need for a dest, and for MountImages empty dest doesn't make sense. So they fit together and complement, so to speak. Implementation is a bit easier too, I think. Do you see cases where it would be useful on a subtree?

bluca replied:
Moved to new directive RootExtensions as discussed

poettering reviewed (commented):
too late for today to do the full review, just some early comments

boucman commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -8026,6 +8046,8 @@ node /org/freedesktop/systemd1/unit/dev_2dsda3_2eswap {
 
     <!--property RootVerity is not documented!-->
 
+    <!--property RootExtensions is not documented!-->
```
please document the new property...

bluca replied:
mmh I'm not entirely sure if we want to document these kind of properties explicitly? I mean, it's a 1:1 mapping to the unit setting after all - name and values coincide

keszybz replied:
Then please mention it in the text. Even something as simple as: "The following properties map 1:1 to corresponding settings in the unit file: RootVerity, RootExtension, ..., see systemd.service(5) for their meaning."

bluca replied:
Ok, done

boucman commented on man/systemd.exec.xml:
```diff
@@ -271,6 +271,25 @@
+
+        <listitem><para>Takes a space-separated list of paths to block device nodes or regular files as argument.
+        For each image located in such paths, a variable overlay of <filename>/usr/</filename> and
+        <filename>/opt/</filename> is set up on the root. Nodes in the image that are leaves in the root (empty
```
Not exactly related but /opt is not documented at all in the file-hierarchy man page. 
So why treat it as a special case ? what is supposed to go in there ?

bluca replied:
It is documented in the FSH 3.0: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s13.html
It's for non-distro-vendor extensions, commonly used for binaries (eg: chrome is installed in /opt). So it's a third-party /usr, so to speak.

boucman replied:
I thought /usr/local was the place for that, but ok...

bluca replied:
The /usr/local hierarchy is for programs built locally and installed by the local system administrator/user

poettering commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+        char ***overlays;
+        size_t extra_last;
+};
+static thread_local struct nftw_cb_data *cb_data;
```
actually, thinking about this, i have the suspicion that avoiding nftw here would be simple enough and much prettier.

we already do that in various places. for example, see clean_posix_shm_internal() in src/shared/clean-ipc.c which recursively descends /dev/shm. I'd argue that code is more readable and probably even shorter than anything we could possibly cook up with nftw().

bluca replied:
Will work on that next as discussed, in the meanwhile pushed with the other requested changes.

bluca replied:
Added a path_breadth_first_visit helper, and the mount_compute_shallow_overlays helper uses it with a local callback.

poettering commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+        check = prefix_roota(cb_data->root_path, &fpath[cb_data->extra_last]);
+        r = access(check, F_OK);
+        if (r == 0) {
+                /* file/dir exists in the root */
```
you need to take the mounts from b_data->mounts into account here... as any earlier layer might have the dir, not just the base...

i.e. ->mounts should probably be a hashmap keyed by the path, returning the dir it shall be bind mounted from.

bluca replied:
uhm the way it's intended to use, as discussed yesterday, is to compare the fully applied view with the extension - ie, (((root + a) + b) + c), so the base will have everything already when this comparison is done

Or am I misunderstanding the point?

poettering commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+                if (!!(tflag == FTW_F) != !!(S_ISREG(st.st_mode)))
+                        return -EINVAL;
+                /* Individual file, or directory that is empty in the root - just overmount */
+                if (tflag == FTW_F || dir_is_empty(check))
```
dir_is_empty() returns ngetaive on error

poettering commented on man/systemd.exec.xml:
```diff
@@ -271,6 +271,25 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RootExtensions=</varname></term>
```
I wonder how RootExtensions= is actually the right name for this? they don't actually extend the root of the hierarchy, after all, but /usr/ + /opt/ only.

Moreover so far we always had two distinct options for images and trees. i.e. in nspawn --directory= and --image=. for services RootImage= and RootDirectory=. It probably makes sense to continue this logic.

Hence using that as blueprint have two options, maybe named ExtensionImage= and ExtensionDirectory= instead?

bluca replied:
Sure thing - but ExtensionImages (plural) since it's a multi-value option. Also I don't have ExtensionDirectories ready - is it ok if it's a follow-up?

poettering reviewed (changes requested):
here's half a review. too late today for more. will look at this furher tomorow.

bluca commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+
+        /* file/dir does not exists in the root - if the parent is empty, just overmount */
+        parent = dirname((char *)check);
+        if (dir_is_empty(parent)) {
```
good catch

bluca commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+        int r;
+
+        if (tflag != FTW_F && tflag != FTW_D)
+                return 0;
```
oops yes absolutely, thanks

bluca commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,112 @@ bool credential_name_valid(const char *s) {
+                return 0;
+
+        /* Ignore /usr and /opt themselves, we only care about children */
+        if (STR_IN_SET(&fpath[cb_data->extra_last], "/usr", "/opt"))
```
ok - same variable name or different?

bluca replied:
Used the same variable for now, and copied parsing logic from sysext. After sysext is merged I'll update to refactor it in a common helper.

keszybz commented on src/core/dbus-execute.c:
```diff
@@ -3452,6 +3453,41 @@ int bus_exec_context_set_transient_property(
+
+                STRV_FOREACH(p, l) {
+                        if (!path_is_absolute(*p) || !path_is_normalized(*p))
+                                return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS, "Invalid %s", name);
```
"Invalid path in %s" ?

keszybz commented on src/core/load-fragment-gperf.gperf.m4:
```diff
@@ -27,6 +27,7 @@ $1.RootImageOptions,                     config_parse_root_image_options,
 $1.RootHash,                             config_parse_exec_root_hash,                 0,                                  offsetof($1, exec_context)
 $1.RootHashSignature,                    config_parse_exec_root_hash_sig,             0,                                  offsetof($1, exec_context)
 $1.RootVerity,                           config_parse_unit_path_printf,               true,                               offsetof($1, exec_context.root_verity)
+$1.ExtensionImages,                       config_parse_namespace_path_strv,            0,                                  offsetof($1, exec_context.extension_images)
```
indentation

keszybz commented on src/core/namespace.c:
```diff
@@ -1019,6 +1024,33 @@ static int mount_images(const MountEntry *m) {
+        if (!escaped_source)
+                return -ENOMEM;
+
+        /* We only support read-only overlays, so no upper nor work directories */
```
nor → or

keszybz commented on src/core/namespace.c:
```diff
@@ -1334,6 +1370,83 @@ static void normalize_mounts(const char *root_directory, MountEntry *mounts, siz
+                size_t *n_mounts,
+                char **error_path) {
+        _cleanup_free_ char **deny_list = NULL;
+        MountEntry *m = NULL;
```
Please move the definition of `m` into the loops below.

keszybz reviewed (commented):
Looks OK in general. I didn't read the details of the setup functions though.

bluca commented on test/test-functions:
```diff
@@ -465,6 +465,35 @@ setup_basic_environment() {
+            veritysetup format $oldinitdir/usr/share/minimal.raw $oldinitdir/usr/share/minimal.verity | \
+                grep '^Root hash:' | cut -f2 | tr -d '\n' > $oldinitdir/usr/share/minimal.roothash
+        )
+    fi
```
Good point, I think it was then at some point I refactored but forgot to update the commit message. Done now.

bluca commented on src/basic/path-util.c:
```diff
@@ -1172,3 +1173,173 @@ bool credential_name_valid(const char *s) {
+        if (path_startswith_strv(current_dir, *cb_data->mounts) != NULL)
+                return 0;
+        if (path_startswith_strv(current_dir, *cb_data->overlays) != NULL)
+                return 0;
```
Yep, not sure what I was thinking, fixed

bluca commented on man/systemd.exec.xml:
```diff
@@ -271,6 +271,25 @@
+      <varlistentry>
+        <term><varname>ExtensionImages=</varname></term>
+
+        <listitem><para>Takes a space-separated list of paths to block device nodes or regular files as argument.
```
Thanks, much better

bluca commented:
> Looks OK in general. I didn't read the details of the setup functions though.

Thanks for the review - yes the path-util.c change is going to get refactored as discussed with Lennart, so no need to review it now. Will hopefully get that done today or tomorrow. Rest is fixed.

bluca commented:
This is ready for review. I changed to do overlayfs only, like sysext, and will follow up in the near future if I need a non-overlay fallback. But this is good to go from my point of view. PTAL!

bluca commented:
@mrc0mmand any idea what those modprobe errors on centos/arch are?

> E: modprobe: FATAL: Module locking.h not found in directory /lib/modules/5.10.15-arch1-1

mrc0mmand commented:
> @mrc0mmand any idea what those modprobe errors on centos/arch are?
> 
> > E: modprobe: FATAL: Module locking.h not found in directory /lib/modules/5.10.15-arch1-1

That's strange, I'll check what's going on.

mrc0mmand commented:
> @mrc0mmand any idea what those modprobe errors on centos/arch are?
> 
> > E: modprobe: FATAL: Module locking.h not found in directory /lib/modules/5.10.15-arch1-1

Ah, I see. The issue is not in the `modprobe` errors (these are harmless), but there's no `VERSION_ID=` field in `/etc/os-release`:

```
+ veritysetup format /var/tmp/systemd-test.wC5CpU/root/usr/share/minimal_1.raw /var/tmp/systemd-test.wC5CpU/root/usr/share/minimal_1.verity
+ grep '^Root hash:'
+ cut -f2
+ tr -d '\n'
+ export initdir=/var/tmp/systemd-test.wC5CpU/app0
+ initdir=/var/tmp/systemd-test.wC5CpU/app0
+ mkdir -p /var/tmp/systemd-test.wC5CpU/app0/usr/lib/extension-release.d /var/tmp/systemd-test.wC5CpU/app0/usr/lib/systemd/system /var/tmp/systemd-test.wC5CpU/app0/opt
+ grep '^ID=' /etc/os-release
+ grep '^VERSION_ID=' /etc/os-release
make: *** [Makefile:2: setup] Error 1
```

```
# cat /etc/os-release 
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://www.archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux
```

bluca commented:
ah of course - missed that, thank you, will fix

bluca commented:
Fixed, and relaxed the check for sysext too, given the same issue will apply.

mrc0mmand commented:
From the intermediate results it looks like both TEST-50 and TEST-58 are passing, nice!

bluca commented:
TEST-58 is using images, so unfortunately it will suffer from the same issues with uevents that affect TEST-50 ( https://github.com/systemd/systemd/issues/17469 ) - as it happened in CentOS 7 :-( Sadly no feedback yet on the kernel patches to fix it

mrc0mmand commented:
> TEST-58 is using images, so unfortunately it will suffer from the same issues with uevents that affect TEST-50 ( #17469 ) - as it happened in CentOS 7 :-( Sadly no feedback yet on the kernel patches to fix it

I see. I restarted the job to see if it's indeed a flake (not that I don't believe you!).

Also, I'm considering implementing a bit ugly hack, which would try to run the known-to-be-flaky tests a couple of times if they fail. That should, hopefully, reduce the flake noise without sacrificing the coverage.

bluca commented:
> > TEST-58 is using images, so unfortunately it will suffer from the same issues with uevents that affect TEST-50 ( #17469 ) - as it happened in CentOS 7 :-( Sadly no feedback yet on the kernel patches to fix it
> 
> I see. I restarted the job to see if it's indeed a flake (not that I don't believe you!).
> 
> Also, I'm considering implementing a bit ugly hack, which would try to run the known-to-be-flaky tests a couple of times if they fail. That should, hopefully, reduce the flake noise without sacrificing the coverage.

Yeah, that would be nice. We should do the same in autopkgtest. The fix is a long way from coming yet

keszybz commented on src/test/test-strv.c:
```diff
@@ -659,6 +659,22 @@ static void test_strv_foreach_pair(void) {
+        STRV_FOREACH_INDEX(check, a, j) {
+                assert_se(i == j);
+                assert_se(streq(*check, input_table_multiple[i++]));
+        }
```
Hmm, I'd prefer to write such iteration as:
```C
for (size_t i = 0; s && s[i]; i++)
   ... s[i]
```
This avoids the need to define two variables and the use of `char**`. The compiler will optimize this just as well or even better.

bluca replied:
Ok, changed and dropped new helper

keszybz commented on man/systemd.exec.xml:
```diff
@@ -433,6 +433,47 @@
+
+        <para>A read-only OverlayFS will be set up on top of <filename>/usr/</filename> and
+        <filename>/opt/</filename> hierarchies from the root. The order in which the images are listed
+        will determine the order in which the overlay is laid down: first to last -> bottom to top.</para>
```
images specified first to last will result in overlayfs layers bottom to top ?

bluca replied:
much better, yes

keszybz commented on src/core/execute.c:
```diff
@@ -5659,6 +5665,19 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
+                LIST_FOREACH(mount_options, o, c->extension_images[i].mount_options)
+                        fprintf(f, ":%s:%s",
+                                partition_designator_to_string(o->partition_designator),
+                                o->options);
```
isn't `strempty()` necessary here?

bluca replied:
whoops, good catch - fixed in a separate commit for MountImages too, since that was borken as well

bluca replied:
actually split the MountImages one out since it's unrelated: https://github.com/systemd/systemd/pull/18728

keszybz commented on src/core/namespace.c:
```diff
@@ -1378,6 +1439,98 @@ static void normalize_mounts(const char *root_directory, MountEntry *mounts, siz
+                        }
+                        if (r == 0) {
+                                /* We hit a symlinked mount point. The entry got rewritten and might
+                                        * point to a very different place now. Let's normalize the changed
```
indentation

keszybz reviewed (commented):
Oh man, a lot of code.

bluca commented:
> Oh man, a lot of code.

...but not as much as it used to be, after the split&merges :-D Thanks for the review, updated, ptal!

bluca commented:
> > Oh man, a lot of code.
> 
> ...but not as much as it used to be, after the split&merges :-D Thanks for the review, updated, ptal!

To try and make it simpler to review, I've split out the setup_namespace() refactor in a separate commit, since it's mostly code moving around.

poettering commented on src/core/namespace.c:
```diff
@@ -1429,6 +1533,112 @@ static bool home_read_only(
+
+}
+
+static int setup_one_tmp_dir(const char *id, const char *prefix, char **path, char **tmp_path) {
```
any reason why these two functions are moved up? they can remain at the end, no? what do you have in mind they shall be used for?

bluca replied:
in the append_extension_images() helper added by the second commit, which is called from setup_namespaces, so it needs to be above

bluca replied:
but now with the /run/systemd/unit-extesions/[1-2-3] it is no longer needed, so this change was removed

poettering commented on src/core/namespace.c:
```diff
@@ -1173,7 +1173,7 @@ static int apply_mount(
 
         case MOUNT_IMAGES:
-                return mount_images(m);
+                return mount_image(m, NULL);
```
hmm, the function only has only one argument further up? rebase mistake? /me is confused

bluca replied:
yep, refactor mistake - I've split the commit that moves/renames from the "new feature commit", and this fell through, fixed now

poettering commented on src/core/load-fragment.c:
```diff
@@ -5117,6 +5117,148 @@ int config_parse_mount_images(
+                }
+
+                s = sresolved;
+                if (s[0] == '-') {
```
hmm, i think this should be parsed out *before* applying unit_full_printf(), i.e. it should not be something a specifier might synthesize. i.e. just swap this and the previous bit

bluca replied:
ok done (it was copy/pasted from the MountImages parsing, will fix it there too separately)

bluca replied:
https://github.com/systemd/systemd/pull/18744

poettering commented on src/core/namespace.c:
```diff
@@ -494,6 +500,15 @@ static int append_protect_system(MountEntry **p, ProtectSystem protect_system, b
+                        return -1;
+                else
+                        return 1;
+        }
```
```c
r = -CMP(a->mode == EXTENSION_IMAGES, b->mode == EXTENSION_IMAGES)
if (r != 0)
        return r;
```

no?

bluca replied:
uh, much nicer yes :-)

poettering commented on src/core/namespace.c:
```diff
@@ -640,7 +655,8 @@ static void drop_outside_root(const char *root_directory, MountEntry *m, size_t
 
-                if (!path_startswith(mount_entry_path(f), root_directory)) {
+                /* ExtensionImages bases are opened in the propagate directory intentionally */
+                if (f->mode != EXTENSION_IMAGES && !path_startswith(mount_entry_path(f), root_directory)) {
```
the debug log message this will generate is misleading, maybe handle this in a separate if block with a debug message of its own

bluca replied:
mmh sorry I don't grok this, the log message is printed in the same situations as before - we are simply skipping it when we process the new type now. What would be the confusion?

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2620,7 +2620,14 @@ static const char *const partition_designator_table[] = {
+                const char *src,
+                const char *dest,
+                const MountOptions *options,
+                const char *host_os_release_id,
```
maybe name this parameters `required_host_os_release_id` or so? i.e. these are the things

poettering commented on src/shared/dissect-image.c:
```diff
@@ -2682,6 +2689,29 @@ int verity_dissect_and_mount(const char *src, const char *dest, const MountOptio
+                        host_os_release_sysext_level,
+                        extension_release);
+                if (r == 0)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Image %s extension-release metadata does not match the root's", dissected_image->image_name);
```
maybe use some recognizable error message for this. ESTALE? EBADSLT? EDOM?

bluca replied:
changed to ESTALE

poettering commented on src/core/namespace.h:
```diff
@@ -93,11 +93,19 @@ struct TemporaryFileSystem {
+
 struct MountImage {
         char *source;
         char *destination;
```
hmm, so you are overloading this structure with a new purpose. which i guess is OK, given that setup_namespace() already is a ridiculous number of parameters and this is close enough. but please add a comment to the `char *destination` line, that clarifies the field is unsigned if MountImageType == MOUNT_IMAGE_OVERLAY.


(and one day we should probably move many of the parameters to setup_namespace() into NamespaceInfo)

poettering commented on src/core/namespace.h:
```diff
@@ -93,11 +93,19 @@ struct TemporaryFileSystem {
 
+typedef enum MountImageType {
+        MOUNT_IMAGE_DISCRETE,
+        MOUNT_IMAGE_OVERLAY,
```
should be named MOUNT_IMAGE_EXTENSION, no? it doesn't do generic overlays after all, only our extensions

bluca replied:
I kinda went with "how it's used" rather than "what it implements", but changed as suggested

poettering commented on src/core/namespace.c:
```diff
@@ -1817,6 +1986,23 @@ int setup_namespace(
+                        hierarchies = strv_new("/usr", "/opt");
+                        if (!hierarchies)
+                                return -ENOMEM;
+                }
```
maybe add a helper for this after all, to os-util.h or so, that gets the list of hierarchies.

poettering commented on src/core/namespace.c:
```diff
@@ -1639,6 +1700,111 @@ static int setup_one_tmp_dir(const char *id, const char *prefix, char **path, ch
+                if (!prefixed_hierarchy)
+                        return -ENOMEM;
+
+                strv_consume(&overlays, TAKE_PTR(prefixed_hierarchy));
```
this can fail ,needs oom check

bluca replied:
uh of course, d'oh

poettering commented on src/core/namespace.c:
```diff
@@ -1639,6 +1700,111 @@ static int setup_one_tmp_dir(const char *id, const char *prefix, char **path, ch
+                strv_consume(&overlays, TAKE_PTR(prefixed_hierarchy));
+        }
+
+        propagate_parent = dirname_malloc(propagate_dir);
```
mm, what does this have to use with popagation? i don't get it? you could just mount this to `/run/systemd/unit-extension/{0,1,2,3,4,5}` or so, no? i.e. do it like for the root image, which we mount to `/run/systemd/unit-root`. Given that mount propagation for the whole of /run is turned off here, you can just take a dir of your own. (And i'd number the mounts, since after all the mount points to overmount have to be created and that way you don't need to clean them up really)

bluca replied:
Didn't want to add a new directory, and that one was already passed through as a unit-private dir. But changed to numbered mount points unbder unit-extension/ as suggested, as it is indeed much nicer.

bluca commented on src/basic/strv.h:
```diff
@@ -127,6 +127,9 @@ bool strv_overlap(char * const *a, char * const *b) _pure_;
         for ((x) = (l), (y) = (x) ? (x+1) : NULL; (x) && *(x) && *(y); (x) += 2, (y) = (x + 1))
 
+#define STRV_FOREACH_INDEX(s, l, i)              \
+        for ((s) = (l), (i) = 0; (s) && *(s); (s)++, (i)++)
```
yep this commit was dropped

bluca commented:
Rebased to fix merge conflicts - PTAL

keszybz commented on src/shared/extension-release.c:
```diff
@@ -77,3 +77,18 @@ int extension_release_validate(
+
+        r = getenv_path_list("SYSTEMD_SYSEXT_HIERARCHIES", ret_hierarchies);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse SYSTEMD_SYSEXT_HIERARCHIES environment variable : %m");
```
Hmm, we generally ignore invalid variable values. Especially if this is used in "shared" code. Shouldn't the error be suppressed here?

bluca replied:
it was copied from the existing implementation verbatim - but you are right, I'll send a follow-up

bluca replied:
https://github.com/systemd/systemd/pull/18796

keszybz reviewed (commented):
I looked at the code again, and don't see any obvious issues.
This has been reviewed a gazillion times now... and even if there's something off, we're more likely to find it by playing with the code.

poettering commented on src/core/namespace.c:
```diff
@@ -1378,6 +1378,110 @@ static void normalize_mounts(const char *root_directory, MountEntry *mounts, siz
+        if (!proc_self_mountinfo) {
+                if (error_path)
+                        *error_path = strdup("/proc/self/mountinfo");
+                return log_debug_errno(errno, "Failed to open /proc/self/mountinfo: %m");
```
hmm, errno might be set due to strdup() failing, instead of the actual error.

bluca replied:
ops, good catch

poettering commented:
hmm, i'll prep a PR for these fix-ups shortly.

bluca commented on src/shared/extension-release.c:
```diff
@@ -77,3 +77,18 @@ int extension_release_validate(
+                *ret_hierarchies = strv_new("/usr", "/opt");
+                if (!*ret_hierarchies)
+                        return -ENOMEM;
+        }
```
Yeah it was copied from another place but I should have adapted it - already a change in the pipeline in https://github.com/systemd/systemd/pull/18797
